articles/ros2_security_contexts.md

```diff
@@ -0,0 +1,234 @@
+
+For nodes with fully qualified context paths, namespacs do not subsequently push the relative sub-folder.
+
+``` xml
+<launch>
+  <group>
+    <push_ros_namespace namespace="foo"/>
+    <node pkg="demo_nodes_cpp" exec="listener" context="/bar"/>
+  </group>
+</launch>
+```
+
+```
+$ tree --dirsfirst contexts/
+contexts/
+└── bar
+    ├── cert.pem
+    ├── governance.p7s
+    ├── identity_ca.cert.pem
+    ├── key.pem
+    ├── permissions_ca.cert.pem
+    └── permissions.p7s
+```
+
+
+## Alternatives
+
+### Context path orthogonal to namespace
+
+An alternative to reusing namespaces to hint the context path could be to completely disassociate the two entirely, treating the context path as it's own unique identifier. However, having to book keep both identifier spaces simulations may introduce to many degrees of freedom that a human could groc or easily introspect via tooling.
+
+#### `<push_ros_namespace namespace="..." context="foo"/>`
+
+TODO: Describe added `context` attribute to `push_ros_namespace` element.
+Keeps pushing contexts close/readable to pushing of namespaces.
+
+#### `<push_ros_context context="foo"/>`
+
+TODO: Describe added `push_ros_context` element.
+Keeps pushing context path independent/flexable from namespaces.
+
+
+## Concerns
+
+### Multiple namespaces per context
+
+For circumstances where users may compose multiple nodes of dissimilar namespaces into a single context, the user must subsequently specify a common fully qualified context path for each node to compose, as the varying different namespaces would not push to a common context. For circumstances where the context path is orthogonal to node namespace, the use of fully qualifying all relevant nodes is could be tedious, but could perhaps could still be parametrized via the use of `<var/>`, and `<arg/>` substation and expansion.
+
+
+### Multiple contexts per process
+
+As before the use of contexts, multiple nodes composed into a single process where each mapped to a septate participant. Each participant subsequently load an security identity and access control credential prevalent to it's respective node. This composition however would inevitably mean that code compiled to node `foo` could access credentials/permissions only trusted only to node `bar`. This consequence of composition could unintendedly subvert the minimal spanning policy as architected by the designer or measured/generated via ROS 2 tooling/IDL.
+
+With the introduction of contexts, it becomes possible to describe the union of access control permission by defining a collection of SROS 2 policy profiles as element within a specific context. This would allow for formal analysis tooling to check for potential violations in information flow control given the composing of nodes at runtime. However, should multiple contexts be used per process, then such security guaranties are again lost. Thus it should be asked whether if multiple contexts per process should even be supported.
```

Comment:
I'm not sure I got this part. Why would the security guaranties be lost if there are multiple contexts per process?

Reply:
Please comment on a461e4c .